### PR TITLE
Implement `inverseSqrt` tests

### PR DESCRIPTION
**Pull request by zoddicus**
**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [x] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [x] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/builtin/inversesqrt.spec.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// Non-linear range generator, biased to towards min.
+function biasedRange(min: Scalar, max: Scalar, steps: Scalar): Array<number> {
+  const { f32_min, f32_max, u32_steps } = unwrapRangeParams(min, max, steps);
+
+  return Array.from(Array(u32_steps).keys()).map(i =>
+    lerp(f32_min, f32_max, Math.pow(lerp(0, 1, i / (u32_steps - 1)), 2))
+  );
+}
```
Comment (zoddicus) [inline src/webgpu/shader/execution/builtin/inversesqrt.spec.ts:70]: All of the range generation code can probably be factored out into a utility file and used by other tests that test a range of programmatically values.

I wanted to get this reviewed before I did that refactoring, and also wasn't sure if I should do that in a separate PR. i.e. land this test PR and then have a follow up that factors out the utility code and update other tests to use it.

**Comment by zoddicus** [comment]
PTAL

File: src/webgpu/shader/execution/builtin/inversesqrt.spec.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    return min;
+  }
+
+  if (min > max) {
+    const temp = min;
+    min = max;
+    max = temp;
+  }
+
+  if (t > 1) {
```
Comment (dneto0) [inline src/webgpu/shader/execution/builtin/inversesqrt.spec.ts:29]: I'd include the end range too, just for potential tidiness.   

      t >= 1

File: src/webgpu/shader/execution/builtin/inversesqrt.spec.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    return max;
+  }
+
+  if (t < 0) {
```
Comment (dneto0) [inline src/webgpu/shader/execution/builtin/inversesqrt.spec.ts:33]: Similarly  `t  <= 0`

File: src/webgpu/shader/execution/builtin/inversesqrt.spec.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    return min;
+  }
+
+  return min + t * (max - min);
```
Comment (dneto0) [inline src/webgpu/shader/execution/builtin/inversesqrt.spec.ts:37]: I wonder if this is good numerically.  For example, if max and min are both large magnitude and close to each other, then you get catastrophic cancellation.
(Subtracting to float numbers that agree to k bits means you lose k bits of precision, generally).

My next thought was to suggest:


       (min * (1-t)  + max * t)

Then I googled "numerically stable linear interpolation" and found someone who has thought about this way more deeply. There is a relatively simple solution that has nice properties.

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0811r2.html 

Please use the suggested algorithm.

File: src/webgpu/shader/execution/builtin/inversesqrt.spec.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// Unwrap Scalar params into numbers and test preconditions
+function unwrapRangeParams(min: Scalar, max: Scalar, steps: Scalar) {
```
Comment (dneto0) [inline src/webgpu/shader/execution/builtin/inversesqrt.spec.ts:41]: What is the steps parameter? The size of the step between min and max, or the number of steps.

I have to read the body to be sure.

File: src/webgpu/shader/execution/builtin/inversesqrt.spec.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+  return Array.from(Array(u32_steps).keys()).map(i => lerp(f32_min, f32_max, i / (u32_steps - 1)));
+}
+
+// Non-linear range generator, biased to towards min.
+function biasedRange(min: Scalar, max: Scalar, steps: Scalar): Array<number> {
+  const { f32_min, f32_max, u32_steps } = unwrapRangeParams(min, max, steps);
+
+  return Array.from(Array(u32_steps).keys()).map(i =>
+    lerp(f32_min, f32_max, Math.pow(lerp(0, 1, i / (u32_steps - 1)), 2))
```
Comment (dneto0) [inline src/webgpu/shader/execution/builtin/inversesqrt.spec.ts:68]: Interesting.  So you square the lerp'd steps, so it's like a up-facing parabola (quadratic) from the origin to (1,1), if I am reading this correctly.
A comment summarizing that would be nice.

File: src/webgpu/shader/execution/builtin/inversesqrt.spec.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+Please read the following guidelines before contributing:
+https://github.com/gpuweb/cts/blob/main/docs/plan_autogen.md
+`
+  )
+  .params(u =>
+    u
+      .combine('storageClass', ['uniform', 'storage_r', 'storage_rw'] as const)
+      .combine('vectorize', [undefined, 2, 3, 4] as const)
+  )
+  .fn(async t => {
```
Comment (dneto0) [inline src/webgpu/shader/execution/builtin/inversesqrt.spec.ts:89]: This is really nice!

**Review by dneto0 — CHANGES_REQUESTED** [review]
I have some nitpicks about comments, and a request for a better (more numerically good) lerp function.  I guess the lerp niceness doesn't matter in this case, but it may get reused where it does matter.

File: src/webgpu/shader/execution/builtin/inversesqrt.spec.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/builtin/inversesqrt.spec.ts:94]: Once this PR lands I will refactor and send out another PR.

**Comment by zoddicus** [comment]
@dneto0 this is ready for another round of review

**Review by dneto0 — APPROVED** [review]
Looks great!

**Pull request merged**